orca-pipelinetemplate/src/main/java/com/netflix/spinnaker/orca/pipelinetemplate/v2schema/V2SchemaExecutionGenerator.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright 2018 Google, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.orca.pipelinetemplate.v2schema;
+
+import com.netflix.spinnaker.orca.pipelinetemplate.TemplatedPipelineRequest;
+import com.netflix.spinnaker.orca.pipelinetemplate.generator.V2ExecutionGenerator;
+import com.netflix.spinnaker.orca.pipelinetemplate.v1schema.TemplateMerge;
+import com.netflix.spinnaker.orca.pipelinetemplate.v1schema.model.NamedHashMap;
+import com.netflix.spinnaker.orca.pipelinetemplate.v1schema.model.TemplateConfiguration;
+import com.netflix.spinnaker.orca.pipelinetemplate.v2schema.model.V2PipelineTemplate;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class V2SchemaExecutionGenerator implements V2ExecutionGenerator {
+
+  @Override
+  public Map<String, Object> generate(V2PipelineTemplate template, TemplateConfiguration configuration, TemplatedPipelineRequest request) {
+    Map<String, Object> pipeline = new HashMap<>();
+    pipeline.put("id", Optional.ofNullable(request.getId()).orElse(Optional.ofNullable(configuration.getPipeline().getPipelineConfigId()).orElse("unknown")));
+    pipeline.put("application", configuration.getPipeline().getApplication());
+    if (request.getExecutionId() != null) {
+      pipeline.put("executionId", request.getExecutionId());
+    }
+    pipeline.put("name", Optional.ofNullable(configuration.getPipeline().getName()).orElse("Unnamed Execution"));
+
+    V2PipelineTemplate.Configuration c = template.getConfiguration();
+    Map<String, Object> concurrentExecutions = (Map) c.get("concurrentExecutions");
+    if (concurrentExecutions == null || concurrentExecutions.isEmpty()) {
+      pipeline.put("limitConcurrent", request.isLimitConcurrent());
+      pipeline.put("keepWaitingPipelines", request.isKeepWaitingPipelines());
+    } else {
+      pipeline.put("limitConcurrent", concurrentExecutions.getOrDefault("limitConcurrent", request.isLimitConcurrent()));
+      pipeline.put("keepWaitingPipelines", concurrentExecutions.getOrDefault("keepWaitingPipelines", request.isKeepWaitingPipelines()));
+    }
+
+    addNotifications(pipeline, template, configuration);
+    addParameters(pipeline, template, configuration);
+    addTriggers(pipeline, template, configuration);
+
+    pipeline.put("templateVariables", configuration.getPipeline().getVariables());
+
+    pipeline.put("stages", template.getStages()
```

Comment:
Is there a way we can cut down on this manual transformation? It will mean additional work whenever the pipeline spec changes.

Reply:
Refactored the v2 template and config models so that we do a lot less extra processing and the template/config definitions are much closer to the models of non-templated pipelines/stages.